src/lightning/data/streaming/functions.py

```diff
@@ -286,3 +287,51 @@ def optimize(
         num_nodes,
         machine,
     )
+
+
+def _listdir(folder: str) -> Tuple[str, List[str]]:
+    return folder, os.listdir(folder)
+
+
+class walk:
+    """This class is an optimized version of os.walk for listing files and folders from cloud filesystem.
+
+    Note: The order of files and folders yielded aren't depth-first anymore due to the asynchronous listing call.
+
+    """
+
+    def __init__(self, folder: str, max_workers: Optional[int] = os.cpu_count()) -> None:
+        self.folders = [folder]
+        self.max_workers = max_workers or 1
+        self.futures: List[concurrent.futures.Future] = []
+
+    def __iter__(self) -> Any:
+        """This function queues the folders to perform listdir across multiple workers."""
+        with concurrent.futures.ThreadPoolExecutor(max_workers=self.max_workers) as executor:
+            while len(self.folders):
+                folder = self.folders.pop(0)
+                future = executor.submit(_listdir, folder)
+                self.futures.append(future)
+
+            while self.futures:
+                for future in concurrent.futures.as_completed(self.futures):
```

Comment:
it is highly discouraged to add to see same list while iterating over it. This won't be reflected since often the iterator is already pre-built. This code could be simplified a lot by using https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future.add_done_callback

Since you don't need a progress bar, you also don't need to manually wait for all of them, just use https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.wait

Reply:
The as_compleleted is while look internally, so it works just fine and keeps the code simpler. In practice, this works quite well.